src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,215 @@
+		if (xmlStart != -1 && xmlEnd != -1) {
+			const cpdXml = stdout.slice(xmlStart, xmlEnd + cpdEnd.length);
+			const cpdJson = xml2js(cpdXml, {compact: false, ignoreDeclaration: true}); 
+
+			const duplications =  cpdJson.elements[0].elements;
+			if (duplications) {
+				violations = this.jsonToRuleResults(duplications);
+			}
+		}
+
+		if (violations.length > 0) {
+			this.logger.trace('Found rule violations.');
+		}
+
+		return violations;
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	private jsonToRuleResults(duplications: any): RuleResult[] {    
+		const ruleResults: RuleResult[] = [];
+
+		for (const duplication of duplications) {
+		
+			const parts = duplication.elements;
+			const codeFragment = parts.find(part => {return part.name === 'codefragment'}).elements[0].cdata;
+			const occurences = parts.filter(part => {return part.name === 'file'});
+			const codeFragmentID = checksum(codeFragment).slice(0,7);
+			// TODO: check for collisions (cpd can return multiple duplications for the same code fragment, so these have the same checksum currently)
+
+			let occCount = 1;
+			for (const occ of occurences) {
+				// create a violation for each occurence of the code fragment
+				const violation: RuleViolation = {
+					line: occ.attributes.line,
+					column: occ.attributes.column,
+					endLine: occ.attributes.endline,
+					endColumn: occ.attributes.endcolumn,
+					ruleName: this.cpdCatalog.rules[0].name,
+					severity: 3,
+					message: `${codeFragmentID}: ${occCount} of ${occurences.length} duplication segments detected. ${duplication.attributes.lines} line(s), ${duplication.attributes.tokens} tokens.`, 
+					category: this.cpdCatalog.categories[0].name,
+					url: 'https://pmd.github.io/latest/pmd_userdocs_cpd.html#refactoring-duplicates',
+				};
+				occCount++;
+
+				// if there are already violations for a file, add to its violations, if not, create a new RuleResult for that file.
+				const ruleResult: RuleResult = ruleResults.find(ruleResult => {return occ.name === ruleResult.fileName})
+				if (ruleResult) {
+					ruleResult.violations.push(violation);
+				} else {
+					ruleResults.push(
+						{
+							engine: CpdEngine.ENGINE_NAME,
+							fileName: occ.attributes.path,
+							violations: [violation]
+						}
+					);
+				}
+			}
+		}
+		return ruleResults;
+	}
+
+	
+	public async init(): Promise<void> {
+		if (this.initialized) {
+			return;
+		}
+		this.logger = await Logger.child(this.getName());
+		this.config = await Controller.getConfig();
+		this.initialized = true;
+		this.logger = await Logger.child(this.getName());
+		this.initialized = true;
+
+		this.cpdCatalog = {
+			rules: [{
+				engine: ENGINE.CPD.valueOf(),
+				sourcepackage: ENGINE.CPD.valueOf(),
+				name: 'copy-paste-detected',
+				description: 'Identify duplicate code blocks.',
+				categories: ['Copy/Paste Detected'],
+				rulesets: [],
+				languages: await this.config.getSupportedLanguages(ENGINE.CPD),
```

Comment:
If someone adds a supported language to config that doesn't have an entry in `languageToPathMap`, we may have some unexpected behavior. We either have to have a default behavior or consider moving languageToPath mapping to config even earlier. What do you think?